src/lib/activity/index.ts

```diff
@@ -0,0 +1,58 @@
+import { backendUrl } from '$lib/config';
+
+// key for storing userId in localStorage
+const USER_ID_KEY = 'userId';
+
+// threshold (in ms) to skip requests of same type
+const THROTTLE_THRESHOLD = 1000;
+
+// store last request timestamp by type
+const lastRequest: Record<string, number> = {};
+
+// track activity to backend by type
+export async function trackActivity(type: string) {
+	// abort if last request too recent
+	const t0 = lastRequest[type] || 0;
+	const t1 = Date.now();
+	if (t1 - t0 < THROTTLE_THRESHOLD) return;
+	lastRequest[type] = t1;
+
+	// POST to /activity endpoint
+	const params = new URLSearchParams({ type, user_id: getUserId() });
```

Comment:
Sanity check - will the type get expanded to `type=chart`? I see that `type` is a string and we only pass `'chart'` into the function?
```ts
trackActivity('chart');
```

Does the TypeScript expand `{name}` to `{name: "value_of_name"}`?

Reply:
yes…
- line 13 means "this function accepts one argument called `type` of type `string`" (a bit confusing since "type" is used in two senses there).
- line 21 is [object property shorthand](https://attacomsian.com/blog/javascript-object-property-shorthand) for
    ```const params = new URLSearchParams({ type: type, user_id: getUserId() });```

So in theory you could call `trackActivity('foo')` and it will try to POST `/activity?type=foo&user_id=…`

…which will fail on the backend. I thought about restricting `type` on `frontend` as well, but I don't love the idea of trying to keep these values in-sync. If someone tries to track an invalid type, the `frontend` will show an error in the console (which can be expanded to show the HTTP response from `backend`), which seems reasonable.

Reply:
OK good, "object property shorthand" is what I was looking for, thanks.

I agree, no sense in duplicating the validation logic on the frontend, because users have no control over the input here. If a backend complains about an invalid type, is the front end programmer's fault. :upside_down_face: